adapter/src/utils/useVerifyLatestUser.js

```diff
@@ -0,0 +1,49 @@
+import {
+    useConfig,
+    useDataQuery /* , clearSensitiveCaches */,
+} from '@dhis2/app-runtime'
+import { useState } from 'react'
+// todo: use app runtime
+import { clearSensitiveCaches } from './clear-sensitive-caches'
+
+const USER_QUERY = {
+    user: {
+        resource: 'me',
+        params: { fields: ['id'] },
+    },
+}
+
+const LATEST_USER_KEY = 'dhis2.latestUser'
+
+/**
+ * This hook is used to clear sensitive caches if a user other than the one
+ * that cached that data logs in
+ * @returns {Object} - { loading: boolean }
+ */
+export function useVerifyLatestUser() {
+    const { pwaEnabled } = useConfig()
+    const [finished, setFinished] = useState(false)
+    const { loading, error } = useDataQuery(USER_QUERY, {
+        onComplete: async data => {
+            const latestUserId = localStorage.getItem(LATEST_USER_KEY)
+            const currentUserId = data.user.id
+            if (currentUserId !== latestUserId) {
+                const cachesCleared = await clearSensitiveCaches()
+                localStorage.setItem(LATEST_USER_KEY, currentUserId)
+
+                if (cachesCleared && pwaEnabled) {
+                    // If this is a PWA app, the app-shell cache will need to
+                    // be restored with a page reload
+                    return window.location.reload()
+                }
```

Comment:
@KaiVandivier I think this works for now, but I don't think it should be necessary... wouldn't this work as-is if we moved the `useVerifyLastUser` usage up one component level, so it completes before the apps shell fetches anything else?

Reply:
Discussed separately, but copied here for others to see: Unfortunately no - if this hook happened above the AppWrapper and below the ServerVersionProvider, the app shell would be missing `api/system/info` in the app shell cache, and would get stuck at the login modal if the app went offline. If the hook happened above ServerVersionProvider, if a user isn't authenticated, then the hook-consuming component wouldn't render the children as-is